debezium-server/debezium-server-eventhubs/src/main/java/io/debezium/server/eventhubs/EventHubsChangeConsumer.java

```diff
@@ -0,0 +1,186 @@
+        maxBatchSize = config.getOptionalValue(PROP_MAX_BATCH_SIZE, Integer.class).orElse(0);
+
+        String finalConnectionString = String.format(CONNECTION_STRING_FORMAT, connectionString, eventHubName);
+
+        try {
+            producer = new EventHubClientBuilder().connectionString(finalConnectionString).buildProducerClient();
+        }
+        catch (Exception e) {
+            throw new DebeziumException(e);
+        }
+
+        LOGGER.info("Using default Event Hubs client for namespace '{}'", producer.getFullyQualifiedNamespace());
+    }
+
+    @PreDestroy
+    void close() {
+        try {
+            producer.close();
+            LOGGER.info("Closed Event Hubs producer client");
+        }
+        catch (Exception e) {
+            LOGGER.warn("Exception while closing Event Hubs producer: {}", e);
+        }
+    }
+
+    @Override
+    public void handleBatch(List<ChangeEvent<Object, Object>> records,
+                            RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        LOGGER.trace("Processing change events...");
+
+        CreateBatchOptions op = new CreateBatchOptions().setPartitionId(partitionID);
+        if (partitionKey != "") {
+            op.setPartitionKey(partitionKey);
+        }
+        if (maxBatchSize.intValue() != 0) {
+            op.setMaximumSizeInBytes(maxBatchSize);
+        }
+
+        EventDataBatch batch = producer.createBatch(op);
+
+        for (ChangeEvent<Object, Object> record : records) {
+            LOGGER.trace("Received event '{}'", record);
+            if (null == record.value()) {
+                continue;
+            }
+            EventData eventData = null;
+
+            if (record.value() instanceof String) {
+                eventData = new EventData((String) record.value());
+            }
+            else if (record.value() instanceof byte[]) {
+                eventData = new EventData(getBytes(record.value()));
+            }
+            try {
+                if (!batch.tryAdd(eventData)) {
+                    LOGGER.warn("event was too large to fit in the batch - {}", record);
+                    continue;
+                }
+            }
+            catch (IllegalArgumentException e) {
+                LOGGER.warn("EventData was null - {}", e.getMessage());
+                continue;
+            }
+            catch (AmqpException e) {
+                LOGGER.warn("EventData is larger than the maximum size of the EventDataBatch - {}", e.getMessage());
+                continue;
+            }
+            catch (Exception e) {
+                LOGGER.warn("Failed to add EventData to batch {}", e.getMessage());
+                continue;
+            }
+
+            // Event Hubs producer only supports "batch"ed sends. Each record is sent as a
+            // separate batch which is then acknowledged/committed
+
+            try {
+                producer.send(batch);
```

Comment:
I think it might be better to send the batch as a single batch and then in a subseqent loop all the records in the batch will be marked as processed. Would not it be more performant? Of course there probably could be duplicates upon restart but IMHO it would be an acceptable tradeoff.
Or you could provide a config option that would affect the behaviour - either as is now (so microbatches) or a batch approach as described above. WDYT?

Reply:
Thanks for the inputs @jpechane. Definitely more performant with a single batch for all the records. I have updated the logic to reflect this